libs/ml/examples/bert_simple_cls_example/main.cpp

```diff
@@ -0,0 +1,307 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "math/tensor.hpp"
+#include "ml/core/graph.hpp"
+#include "ml/layers/fully_connected.hpp"
+#include "ml/layers/normalisation/layer_norm.hpp"
+#include "ml/layers/self_attention_encoder.hpp"
+#include "ml/ops/add.hpp"
+#include "ml/ops/embeddings.hpp"
+#include "ml/ops/loss_functions/cross_entropy_loss.hpp"
+#include "ml/ops/slice.hpp"
+#include "ml/optimisation/adam_optimiser.hpp"
+
+#include <iostream>
+#include <string>
+
+using namespace fetch::ml::ops;
+using namespace fetch::ml::layers;
+
+using DataType   = float;
+using TensorType = fetch::math::Tensor<DataType>;
+using SizeType   = typename TensorType::SizeType;
+using SizeVector = typename TensorType::SizeVector;
+
+using GraphType     = typename fetch::ml::Graph<TensorType>;
+using StateDictType = typename fetch::ml::StateDict<TensorType>;
+using OptimiserType = typename fetch::ml::optimisers::AdamOptimiser<TensorType>;
+
+using RegType         = fetch::ml::RegularisationType;
+using WeightsInitType = fetch::ml::ops::WeightsInitialisation;
+using ActivationType  = fetch::ml::details::ActivationType;
+
+struct BERTConfig
+{
+  // the default config is for bert base uncased pretrained model
+  SizeType n_encoder_layers  = 12u;
+  SizeType max_seq_len       = 512u;
+  SizeType model_dims        = 768u;
+  SizeType n_heads           = 12u;
+  SizeType ff_dims           = 3072u;
+  SizeType vocab_size        = 30522u;
+  SizeType segment_size      = 2u;
+  DataType epsilon           = static_cast<DataType>(1e-12);
+  DataType dropout_keep_prob = static_cast<DataType>(1);
+};
+
+struct BERTInterface
+{
```

Comment:
this looks like a utility function that you re-use in both the load/save example and here. move it to ml/examples/utilities/bert/utilities.hpp

Reply:
oh there are a bunch of these functions tho. things such as read tensor from files, evaluate graphs, BERTConfig, BERTInterface etc.